redbot/core/commands/dpy_reimplements.py

```diff
@@ -0,0 +1,237 @@
+
+    return decorator
+
+
+def guild_only() -> SameObjectDecorator:
+    def pred(ctx: Context) -> bool:
+        if ctx.guild is None:
+            raise dpy_commands.NoPrivateMessage(
+                _("This command cannot be used in private messages.")
+            )
+        return True
+
+    return check(pred)
+
+
+def dm_only() -> SameObjectDecorator:
+    def pred(ctx: Context) -> bool:
+        if ctx.guild is None:
+            raise dpy_commands.PrivateMessageOnly(
+                _("This command can only be used in private messages.")
+            )
+        return True
+
+    return check(pred)
+
+
+def is_nsfw() -> SameObjectDecorator:
+    def pred(ctx: Context) -> bool:
+        channel = ctx.channel
+        if ctx.guild is not None:
+            if isinstance(channel, discord.TextChannel):
+                if not channel.is_nsfw():
+                    raise dpy_commands.NSFWChannelRequired(channel)
+        return True
+
+    return check(pred)
+
+
+def has_role(item: Union[str, int]) -> SameObjectDecorator:
+    def predicate(ctx: Context) -> bool:
+        if ctx.guild is None:
+            raise dpy_commands.NoPrivateMessage(
+                _("This command cannot be used in private messages.")
+            )
+
+        assert isinstance(ctx.author, discord.Member)
+
+        if isinstance(item, int):
+            role = discord.utils.get(ctx.author.roles, id=item)
+        else:
+            role = discord.utils.get(ctx.author.roles, name=item)
+
+        if not role:
+            raise dpy_commands.MissingRole(item)
+
+        return True
+
+    return check(predicate)
+
+
+def has_any_role(*items: Union[int, str]) -> SameObjectDecorator:
+    def predicate(ctx: Context) -> bool:
+        if not ctx.guild:
+            raise dpy_commands.NoPrivateMessage(
+                _("This command cannot be used in private messages.")
+            )
+
+        assert isinstance(ctx.author, discord.Member)
+        getter = functools.partial(discord.utils.get, ctx.author.roles)
+
+        for item in items:
+            if isinstance(item, int):
+                if getter(id=item):
+                    break
+            else:
+                if getter(name=item):
+                    break
+        else:
+            raise dpy_commands.MissingAnyRole(items)
+
+        return True
+
+    return check(predicate)
+
+
+def bot_has_role(item: Union[str, int]) -> SameObjectDecorator:
+    def predicate(ctx: Context) -> bool:
+        if ctx.guild is None:
+            raise dpy_commands.NoPrivateMessage(
+                _("This command cannot be used in private messages.")
+            )
+
+        assert isinstance(ctx.me, discord.Member)
+
+        if isinstance(item, int):
+            role = discord.utils.get(ctx.me.roles, id=item)
+        else:
+            role = discord.utils.get(ctx.me.roles, name=item)
+
+        if not role:
+            raise dpy_commands.MissingRole(item)
```

Comment:
Implementation in d.py seems to have this instead:
```suggestion
            raise dpy_commands.BotMissingRole(item)
```